source_code/main_mcu/src/LOGIC/logic_totp.c

```diff
@@ -0,0 +1,174 @@
+                {
+                    // In this case, there are minimum 5 bits in the current byte at decoded_key[decoded_index] remaining.
+                    // So we can shift the index by (shift_offset - logic_totp_base32_bits) places. For the first character of encoded_key this would be for example
+                    // (shift_offset - logic_totp_base32_bits) = (8 - 5) = 3, so index is shifted 3 places to the left.
+                    // shift_offset is then set to this value for the next character of the encoded_key.
+                    decoded_key[decoded_index] |= j << (shift_offset - LOGIC_TOTP_BASE32_BITS);
+                    shift_offset -= LOGIC_TOTP_BASE32_BITS;
+                }
+                else
+                {
+                    // In this case, we have to split up the 5 bits of the index between the remaining bits of decoded_key[decoded_index]
+                    // and the first bits of decoded_key[decoded_index + 1]. Therefore we first have to shift the index by 5 - shift_offset to the right.
+                    // That means, we have shift_offset (which is also the number of remaining bits in the current byte) bits from the index remaining to put into the current byte.
+                    // Then we need to calculate the new shift_offset, which is basically the old shift_offset + 3 (As already described, the shift_offset is also
+                    // the number of remaining bits in the byte. As we have in the first step used this number of bits from the index, there are 5 - shift_offset bits from the
+                    // index remaining, that have to be put in the next byte. To get the number places the remaining bits have to be shifted in the next byte, we have to subtract
+                    // shift_offset from the number of bits in a byte. If we rearrange the order in that term, we get shift_offset = shift_offset + 8 - 5 = shift_offset + 3)
+                    // As an example for the second character, shift_offset is 3. So we take the first 3 of the 5 index bits and put them into current byte. Then we have 2 bits remaining.
+                    // Those bits need to be shifted then by 6 places to the left, to be at the beginning of the next byte.
+                    decoded_key[decoded_index] |= j >> (LOGIC_TOTP_BASE32_BITS - shift_offset);
+                    shift_offset += LOGIC_TOTP_BITS_PER_BYTE - LOGIC_TOTP_BASE32_BITS;
+                    decoded_index++;
+                    decoded_key[decoded_index] |= j << (shift_offset);
+                }
+            }
+        }
+    }
+}
+
+
+/*! \fn     logic_totp(char* totp_token, const char* key)
+*   \brief  Calculate TOTP Token using given key
+*   \param  calculated totp_token, char[6] expected
+*   \param  Base32 encoded key that is used to calculate totp_token, char[16] expectd
+*   \return void
+*/
+void logic_totp(char* totp_token, const char* key)
+{
+    char hash[20];
+    char decoded_key[10];
```

Comment:
I think you need to support at least a key of 20 bytes. I have a bunch of 2FA TOTP secrets from different services and I have yet to see one different than 32 characters which is 20 bytes decoded right? (20 bytes * 8 / 5)
Does it say anywhere that keys cannot be greater than 20 bytes? It probably would not make a difference in the strength of the sha1 but I don't think I saw anything int he RFC about max key length? But I might have missed it?

Reply:
Already noticed, that the assumption, that the key is always 10 byte decoded is wrong. I will change that. It does not look like there is really a maximum key defined. So it could be, that you can have a longer key, but it looks like 20 bytes is the longest key, that is commonly used.